flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartPipelinedRegionStrategy.java

```diff
@@ -0,0 +1,210 @@
+	 */
+	public RestartPipelinedRegionStrategy(FailoverTopology topology) {
+		this.topology = checkNotNull(topology);
+		this.regions = new HashMap<>();
+
+		// build regions based on the given topology
+		LOG.info("Start building failover regions.");
+		buildFailoverRegions();
+	}
+
+	// ------------------------------------------------------------------------
+	//  region building
+	// ------------------------------------------------------------------------
+
+	private void buildFailoverRegions() {
+		// currently we let a job with co-location constraints fail as one region
+		// putting co-located vertices in the same region with each other can be a future improvement
+		if (topology.containsCoLocationConstraints()) {
+			buildOneRegionForAllVertices();
+			return;
+		}
+
+		// we use the map (list -> null) to imitate an IdentityHashSet (which does not exist)
+		// this helps to optimize the building performance as it uses reference equality
+		final IdentityHashMap<FailoverVertex, HashSet<FailoverVertex>> vertexToRegion = new IdentityHashMap<>();
+
+		// iterate all the vertices which are topologically sorted
+		for (FailoverVertex vertex : topology.getFailoverVertices()) {
+			HashSet<FailoverVertex> currentRegion = null;
+			for (FailoverEdge inputEdge : vertex.getInputEdges()) {
+				if (inputEdge.getResultPartitionType().isPipelined()) {
+					final FailoverVertex producerVertex = inputEdge.getSourceVertex();
+					final HashSet<FailoverVertex> producerRegion = vertexToRegion.get(producerVertex);
+					if (currentRegion == null) {
+						// use producer region as current region and add current vertex to it
+						if (producerRegion == null) {
+							throw new IllegalStateException("Producer task " + producerVertex.getExecutionVertexName()
+								+ " failover region is null while calculating failover region for the consumer task "
+								+ vertex.getExecutionVertexName() + ". This should be a failover region building bug.");
+						}
+						currentRegion = producerRegion;
+						currentRegion.add(vertex);
+					} else {
+						// check if it is the same as the producer region, if so skip the merge
+						// this check can significantly reduce compute complexity in All-to-All PIPELINED edge case
+						if (currentRegion != producerRegion) {
+							// merge current region and producer region
+							for (FailoverVertex v : producerRegion) {
+								vertexToRegion.put(v, currentRegion);
+							}
+							currentRegion.addAll(producerRegion);
+						}
+					}
+				}
+			}
+
+			// generate a new region for a vertex which has no pipelined input
+			if (currentRegion == null) {
+				currentRegion = new HashSet<>(1);
+				currentRegion.add(vertex);
+			}
+			vertexToRegion.put(vertex, currentRegion);
+		}
+
+		// find out all the distinct regions
+		final IdentityHashMap<HashSet<FailoverVertex>, Object> distinctRegions = new IdentityHashMap<>();
+		for (HashSet<FailoverVertex> regionVertices : vertexToRegion.values()) {
```

Comment:
this step is unnecessary; you can replace any use of `distinctRegions` with `regionVertices.values()`

Reply:
of course